### PR TITLE
fix(api): use timing-safe compare for x-probe-secret (#3803)

### LINKED ISSUE #3803: seed-contract-probe uses non-timing-safe string comparison for RELAY_SHARED_SECRET — timing oracle on a cross-service credential
## Summary

`api/seed-contract-probe.ts` authenticates the `x-probe-secret` header using JavaScript's `!==` operator rather than a timing-safe comparison. Every other auth check in the codebase uses `crypto.subtle`-based HMAC or constant-time comparison. This one outlier creates a timing oracle that can be used to recover the `RELAY_SHARED_SECRET` byte-by-byte.

## Evidence

**`api/seed-contract-probe.ts` lines ~225:**
```typescript
const secret = req.headers.get('x-probe-secret');
const expected = process.env.RELAY_SHARED_SECRET;
if (!expected) return jsonResponse({ error: 'not-configured' }, 503, cors);
if (secret !== expected) return jsonResponse({ error: 'unauthorized' }, 401, cors);
```

**Contrast with the rest of the codebase (`api/cache-purge.js`, `server/_shared/internal-auth.ts`):**
```javascript
// Correct pattern used everywhere else:
const isValid = await timingSafeEqual(provided, expected);
```

## Why this matters

`RELAY_SHARED_SECRET` is used across multiple internal services (Vercel functions, Railway relay, probe). Leaking this secret via timing oracle grants access to all auth-gated relay endpoints. The timing difference between an early string-compare exit (wrong first byte) and a full comparison (correct prefix) is measurable over a network with repeated sampling.

## Attack or failure scenario

1. Attacker sends requests to `/api/seed-contract-probe` with varying `x-probe-secret` values.
2. Measures response time distribution for each candidate byte at each position.
3. Recovers the secret prefix statistically over ~1000 requests per byte.
4. Uses recovered `RELAY_SHARED_SECRET` to authenticate against all relay endpoints.

(This attack is practical when the relay is low-latency and the secret is short. It's less practical for long secrets on high-jitter connections, but the vulnerability is architectural.)

## Root cause

Inconsistent security review across API endpoints. The timing-safe helper exists and is used elsewhere; this endpoint was not reviewed against the same standard.

## Recommended fix

Replace the direct string comparison with the same timing-safe helper used throughout the rest of the codebase:
```typescript
import { timingSafeEqual } from '../server/_shared/internal-auth.js';

const isValid = await timingSafeEqual(secret ?? '', expected);
if (!isValid) return jsonResponse({ error: 'unauthorized' }, 401, cors);
```

## Acceptance criteria

- [ ] `api/seed-contract-probe.ts` uses a constant-time comparison for the probe secret.
- [ ] No direct `===` or `!==` string comparison exists for any secret/token value across all `api/` files.

## Suggested labels
security, area: API

## Priority
P2

## Severity
Medium — timing oracle against a shared secret used across multiple services.

## Confidence
Confirmed — contrast with `api/cache-purge.js` and `server/_shared/internal-auth.ts` verified.

### PR DESCRIPTION
**Pull request by koala73**
## Summary

Per issue #3803, \`api/seed-contract-probe.ts\` used \`secret !== expected\` for the \`x-probe-secret\` header — a timing oracle on \`RELAY_SHARED_SECRET\`, which is shared across Vercel functions, the Railway relay, and probe endpoints. Every other internal-auth path in the codebase uses the \`timingSafeEqual\` helper from \`server/_shared/internal-auth.ts\`.

Same template as the #3705 and #3756 fixes shipped earlier today: **bring the outlier site in line with the safer convention the rest of the codebase already uses**. The helper exists, the cost is one import + a one-line call change.

## Change

1. **Export \`timingSafeEqual\`** from \`server/_shared/internal-auth.ts\` (was module-private). JSDoc explains the use case: callers whose secret comes in via a non-\`Authorization\` header (\`x-probe-secret\`, etc.) need the primitive, not the Bearer-token-shaped \`authenticateInternalRequest\`.

2. **\`api/seed-contract-probe.ts\`**: import + replace \`!==\` with \`await timingSafeEqual(secret, expected)\` (with a \`?? ''\` guard for missing header).

3. **Regression test** (\`tests/no-non-timing-safe-secret-compare.test.mts\`):
   - Walks all \`api/*.{ts,js,mjs,cjs}\` files (skipping tests and vendored dirs).
   - Greps for \`(secret|token|bearer|sharedSecret|...) (===|!==) (process.env.* | expected*)\` — the shape of a timing-leaky secret comparison.
   - Comment-stripped to avoid false-positives on doc comments.
   - Allowlist exists for future documented exceptions (currently empty).
   - A second targeted test asserts \`seed-contract-probe.ts\` specifically uses the helper, so this exact regression can't reappear silently.

## Test plan

- [x] 2/2 new tests pass (\`npx tsx --test tests/no-non-timing-safe-secret-compare.test.mts\`)
- [x] \`npm run typecheck\` clean
- [x] Manually verified the source-grep regex catches the pre-fix shape — replacing the new call with the old \`!==\` makes test #1 fail with the right file in the violations list.

## Severity context

The reporter rated this Medium-P2. Practical attack against this specific endpoint is borderline (internal-only ops probes, high network jitter dominates per-byte timing differences for a 32-char secret), but the fix is one line and brings the file in line with the codebase's existing pattern — no good reason not to do it.

Closes #3803

### CODE REVIEW

File: tests/no-non-timing-safe-secret-compare.test.mts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    // Match patterns like:
+    //   secret !== expected
+    //   token === process.env.FOO
+    //   sharedSecret !== process.env.RELAY_SHARED_SECRET
+    // The match is local-variable-name → comparison → env-var-ish RHS.
+    const varAlternation = SECRET_VARS.join('|');
+    const pattern = new RegExp(
+      `\\b(?:${varAlternation})\\b\\s*(?:!==|===)\\s*(?:process\\.env\\.|expected\\b|EXPECTED\\b)`,
+      'i',
+    );
```
[the remark this answers is not in the document]
Comment (koala73) [reply tests/no-non-timing-safe-secret-compare.test.mts]: Good catch — fixed in commit cad01d1bd. Applied your suggested forward + reverse alternation. Also added a meta-test that exercises the regex against 10 representative strings (3 forward, 3 reverse, 4 negative) so the coverage is pinned: a future 'simplification' that drops the reverse arm will fail the meta-test instead of silently regressing.
